### PR TITLE
Fix SAML SSO plugin redirect URL

### PR DESCRIPTION
**Pull request by Luis-3M**
### Description

This PR fixes the issue #6427 -> SAML request must be appended to an IdP URL as a query param with an ampersand, if the URL already contains a question mark, as opposed to always assume that IdP URLs don't have any query params.
Google's IdP URL for instance looks like this: `https://accounts.google.com/o/saml2/idp?idpid=<ID>`, therefore the expected redirect URL would be `https://accounts.google.com/o/saml2/idp?idpid=<ID>&SAMLRequest=<SAMLRequest>`

This code change is backwards compatible with the current behaviour.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Screenshot 2022-06-14 at 14 02 28

Screenshot 2022-06-14 at 14 01 48

### CODE REVIEW

**Comment by nvazquez** [comment]
@Luis-3M @neogismm it seems we have 2 PRs solving the same issue, can you work on unifying them? (#6451)

**Comment by Luis-3M** [comment]
@nvazquez Yep happy to close mine if @neogismm keeps #6451 open 👍

**Comment by neogismm** [comment]
Hi @Luis-3M thanks for the PR. I was stuck on how to write the tests for the same and looking at your code helps me to learn.

**Comment by Luis-3M** [comment]
No worries at all @neogismm - thank you.

**Comment by Luis-3M** [comment]
@nvazquez we will keep this one open.

File: plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/saml/SAMLUtils.java [hunk 1]
```diff
@@ -150,7 +150,7 @@ public static String buildAuthnRequestUrl(final String authnId, final SAMLProvid
             if (spMetadata.getKeyPair() != null) {
                 privateKey = spMetadata.getKeyPair().getPrivate();
             }
-            redirectUrl = idpMetadata.getSsoUrl() + "?" + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(authnRequest), privateKey, signatureAlgorithm);
+            redirectUrl = idpMetadata.getSsoUrl().contains("?") ? idpMetadata.getSsoUrl() + "&" + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(authnRequest), privateKey, signatureAlgorithm) : idpMetadata.getSsoUrl() + "?" + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(authnRequest), privateKey, signatureAlgorithm);
```
Comment (DaanHoogland) [inline plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/saml/SAMLUtils.java:153]: for readibility :
```suggestion
            String appendOperator = idpMetadata.getSsoUrl().contains("?") ? "&" : "?";
            redirectUrl = idpMetadata.getSsoUrl()
                 + appendOperator
                 + SAMLUtils.generateSAMLRequestSignature("SAMLRequest="
                     + SAMLUtils.encodeSAMLRequest(authnRequest), privateKey, signatureAlgorithm);
```

Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/main/java/org/apache/cloudstack/saml/SAMLUtils.java]: Addressed https://github.com/apache/cloudstack/pull/6457/commits/749dca23b7fe86c4dbf811c1afe14ef5f444f413

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java [hunk 2]
```diff
@@ -60,6 +62,32 @@ public void testBuildAuthnRequestObject() throws Exception {
...
+    public void testBuildAuthnRequestUrlWithQueryParam() throws Exception {
+        String consumerUrl = "http://someurl.com";
+        String idpUrl = "http://idp.domain.example?idpid=CX1298373";
+        String spId = "cloudstack";
+        String authnId = SAMLUtils.generateSecureRandomId();
+        DefaultBootstrap.bootstrap();
+        AuthnRequest req = SAMLUtils.buildAuthnRequestObject(authnId, spId, idpUrl, consumerUrl);
+        String appendOperator = idpUrl.contains("?") ? "&" : "?";
+        String redirectUrl = idpUrl + appendOperator + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(req), null, SAML2AuthManager.SAMLSignatureAlgorithm.value());
+        assertEquals(redirectUrl, idpUrl + "&" + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(req), null, SAML2AuthManager.SAMLSignatureAlgorithm.value()));
```
Comment (DaanHoogland) [inline plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:88]: same here, this does not seem to test and ACS code, even thought it calls `SAMLUtils.buildAuthnRequestObject(..)`, it seems to take its output at face value.

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java [hunk 3]
```diff
@@ -60,6 +62,32 @@ public void testBuildAuthnRequestObject() throws Exception {
...
+    public void testBuildAuthnRequestUrlWithoutQueryParam() throws Exception {
+        String consumerUrl = "http://someurl.com";
+        String idpUrl = "http://idp.domain.example";
+        String spId = "cloudstack";
+        String authnId = SAMLUtils.generateSecureRandomId();
+        DefaultBootstrap.bootstrap();
+        AuthnRequest req = SAMLUtils.buildAuthnRequestObject(authnId, spId, idpUrl, consumerUrl);
+        String appendOperator = idpUrl.contains("?") ? "&" : "?";
+        String redirectUrl = idpUrl + appendOperator + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(req), null, SAML2AuthManager.SAMLSignatureAlgorithm.value());
+        assertEquals(redirectUrl, idpUrl + "?" + SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(req), null, SAML2AuthManager.SAMLSignatureAlgorithm.value()));
```
[the remark this answers is not in the document]
Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:75]: @DaanHoogland These two tests are not much different from `testBuildAuthnRequestObject`, except that each one does an assertion on the two possible results of the redirectUrl, based on the code change added in this PR.
Thoughts on how should we implement these tests ?

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:75]: Not sure @Luis-3M , but now they are not testing anything in the production code except maybe consistence of the result of `SAMLUtils.generateSAMLRequestSignature("SAMLRequest=" + SAMLUtils.encodeSAMLRequest(req), null, SAML2AuthManager.SAMLSignatureAlgorithm.value())`, as you are calling it twice and coparing the results of both calls concatenated to something that can be read from this code to be the same. If this function breaks for some reason this test is still going to pass.
`idpUrl` is the same and `appendOperator` is predictable from what you make `idpUrl`.
Depending on what you would test you would hard code the expected result and let the production code produce the same for comparison. I am not sure what the exact format of an SML url would look like and I suppose there would be some part that is unpredictable (e.g. a one time token) so whether and how to test this. I would remove this test if we cannot hardcode anything, as it is giving us false security this way.

[the remark this answers is not in the document]
Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:75]: > I suppose there would be some part that is unpredictable (e.g. a one time token) so whether and how to test this
True we can not predict the value of the SAMLRequest.

Based on the original issue, we know that the result of the `buildAuthnRequestUrl` method must have at least one query param: `SAMLRequest`. The original code was simply appending this same query param with `?`, therefore if the IDP URL already contains at least one query param, then `SAMLRequest` would actually be missing as seen here:

image

Think that we should at least test the existence of this and other parameters already part of the IDP URL, if any, as well as some basic URL components.

Let me know your thoughts @DaanHoogland .

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java [hunk 4]
```diff
@@ -60,6 +64,68 @@ public void testBuildAuthnRequestObject() throws Exception {
...
+        SAMLProviderMetadata idpMetadata = new SAMLProviderMetadata();
+        idpMetadata.setSsoUrl(idpUrl);
+        idpMetadata.setEntityId(idpId);
+
+        URI redirectUrl = new URI(SAMLUtils.buildAuthnRequestUrl(authnId, spMetadata, idpMetadata, SAML2AuthManager.SAMLSignatureAlgorithm.value()));
+        assertThat(redirectUrl).hasScheme(urlScheme);
+        assertEquals(urlScheme, redirectUrl.getScheme());
+        assertThat(redirectUrl).hasHost(idpDomain);
+        assertEquals(idpDomain, redirectUrl.getHost());
+        assertThat(redirectUrl).hasParameter("SAMLRequest");
```
Comment (DaanHoogland) [inline plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:94]:
```suggestion
        assertThat(redirectUrl).hasScheme(urlScheme).hasHost(idpDomain).hasParameter("SAMLRequest");
        assertEquals(urlScheme, redirectUrl.getScheme());
        assertEquals(idpDomain, redirectUrl.getHost());
```

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java [hunk 5]
```diff
@@ -60,6 +64,68 @@ public void testBuildAuthnRequestObject() throws Exception {
...
+        idpMetadata.setSsoUrl(idpUrl);
+        idpMetadata.setEntityId(idpId);
+
+        URI redirectUrl = new URI(SAMLUtils.buildAuthnRequestUrl(authnId, spMetadata, idpMetadata, SAML2AuthManager.SAMLSignatureAlgorithm.value()));
+        assertThat(redirectUrl).hasScheme(urlScheme);
+        assertEquals(urlScheme, redirectUrl.getScheme());
+        assertThat(redirectUrl).hasHost(idpDomain);
+        assertEquals(idpDomain, redirectUrl.getHost());
+        assertThat(redirectUrl).hasParameter("idpid");
+        assertThat(redirectUrl).hasParameter("SAMLRequest");
```
Comment (DaanHoogland) [inline plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java:126]:
```suggestion
        assertThat(redirectUrl).hasScheme(urlScheme).hasHost(idpDomain).hasParameter("idpid").hasParameter("SAMLRequest");
        assertEquals(urlScheme, redirectUrl.getScheme());
        assertEquals(idpDomain, redirectUrl.getHost());
```

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java]: your tests look much cleaner now @Luis-3M , and they seem to actually test the return of the production method, thanks a lot. We can always improve later if we feel we should assert something else.

**Review by DaanHoogland — APPROVED** [review]

**Comment by harikrishna-patnala** [comment]
@Luis-3M Code LGTM. If you can please address one code smell as suggested by Daan, that will look pefect.

**Review by harikrishna-patnala — APPROVED** [review]

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java (hunk 5 above)
Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java]: Addressed https://github.com/apache/cloudstack/pull/6457/commits/2deaefb8d7120b9ee48adebdd18b6fd28c7969b3

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java (hunk 4 above)
Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java]: Addressed https://github.com/apache/cloudstack/pull/6457/commits/2deaefb8d7120b9ee48adebdd18b6fd28c7969b3

**Comment by Luis-3M** [comment]
Thanks for your input @DaanHoogland @harikrishna-patnala .
Suggested changes have been addressed.

File: plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Luis-3M) [reply plugins/user-authenticators/saml2/src/test/java/org/apache/cloudstack/SAMLUtilsTest.java]: No worries at all @DaanHoogland . Thank you for your input on this.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Review by nvazquez — APPROVED** [review]

**Pull request merged**